service-centrifugo/src/main/java/info/bitrich/xchangestream/service/centrifugo/CentrifugoStreamingService.java

```diff
@@ -0,0 +1,132 @@
+import org.coindirect.centrifuge.java.listener.DataMessageListener;
+import org.coindirect.centrifuge.java.message.DataMessage;
+import org.coindirect.centrifuge.java.subscription.SubscriptionRequest;
+import org.coindirect.centrifuge.java.subscription.UnsubscribeRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+public class CentrifugoStreamingService implements DataMessageListener {
+    private static final Logger LOG = LoggerFactory.getLogger(CentrifugoStreamingService.class);
+
+    protected final Centrifugo centrifugo;
+    private State state = State.DISCONNECTED;
+
+    private Map<String, List<ObservableEmitter<String>>> observers = new HashMap<>();
+
+    private enum State {
+        CONNECTING, CONNECTED, DISCONNECTED
+    }
+
+    public CentrifugoStreamingService(CentrifugoToken centrifugoToken) {
+        LOG.debug("Building with "+centrifugoToken);
+        centrifugo = new Centrifugo.Builder(centrifugoToken.getWsUrl())
+                .setUser(new User(centrifugoToken.user, centrifugoToken.client))
+                .setToken(new Token(centrifugoToken.token, centrifugoToken.timestamp))
+                .build();
+    }
+
+
+    public Completable connect() {
+        return Completable.create(e -> {
+            centrifugo.setConnectionListener(new ConnectionListener() {
+                @Override
+                public void onWebSocketOpen() {
+                    LOG.debug("Centrifugo onWebSocketOpen ");
+                    state = State.CONNECTING;
+                }
+
+                @Override
+                public void onConnected() {
+                    LOG.debug("Centrifugo onConnected");
+                    state = State.CONNECTED;
+                    e.onComplete();
+                }
+
+                @Override
+                public void onDisconnected(int i, String s, boolean b) {
+                    LOG.debug("Centrifugo onDisconnected");
+                    state = State.DISCONNECTED;
+                }
+            });
+
+            centrifugo.setDataMessageListener(this);
+
+            centrifugo.connect();
+        });
+    }
+
+    public Completable disconnect() {
+        return Completable.create(completable -> {
+            LOG.info("Disconnecting Centrifugo");
+            centrifugo.disconnect();
+            completable.onComplete();
+        });
+    }
+
+    public Observable<String> subscribeChannel(String channelName) {
+        LOG.info("Subscribing to channel {}.", channelName);
+        Observable<String> observable = Observable.<String>create(e -> {
+            if (state != State.CONNECTED) {
+                LOG.error("Centrifugo not connected.");
+                e.onError(new NotConnectedException());
+                return;
+            }
+
+            LOG.debug("Sending subscribe for {}", channelName);
+            centrifugo.subscribe(new SubscriptionRequest(channelName));
```

Comment:
Will there be any issues if you end up subscribing to the same channel multiple times?